crates/mun_compiler/src/driver.rs

```diff
@@ -0,0 +1,157 @@
+//! `Driver` is a stateful compiler frontend that enables incremental compilation by retaining state
+//! from previous compilation.
+
+use crate::{
+    db::CompilerDatabase,
+    diagnostics::{diagnostics, Emit},
+    PathOrInline,
+};
+use mun_codegen::IrDatabase;
+use mun_hir::{FileId, RelativePathBuf, SourceDatabase, SourceRoot, SourceRootId};
+use mun_target::spec::Target;
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+mod config;
+
+pub use self::config::Config;
+use mun_errors::{Diagnostic, Level};
+use termcolor::WriteColor;
+
+pub const WORKSPACE: SourceRootId = SourceRootId(0);
+
+#[derive(Debug)]
+pub struct Driver {
+    db: CompilerDatabase,
+    out_dir: Option<PathBuf>,
+}
+
+impl Driver {
+    /// Construct a driver with a specific configuration.
+    pub fn with_config(config: Config) -> Self {
+        let mut driver = Driver {
+            db: CompilerDatabase::new(),
+            out_dir: None,
+        };
+
+        // Move relevant configuration into the database
+        driver.db.set_target(config.target);
```

Comment:
Would a single `set_config` be an option? That would simplify adding options and reduce *failure points*.

Reply:
Do you mean on the `Driver` or on the `CompilerDatabase`? There are two considerations

1. The `Config` can contain options for both and/or either the `Driver` and the `Database`.
2. Constructing a `Driver` or a `CompilerDatabase` without configuration doesnt make sense.

Reply:
The compiler database. Now you need to manually set all of the different config fields.